src/python/gudhi/clustering/_tomato.cc

```diff
@@ -224,54 +255,54 @@ auto merge(py::array_t<Cluster_index, py::array::c_style> children, Cluster_inde
     ds_bas[i].name = -1;
   }
   for (Cluster_index m = 0; m < n_leaves - n_final; ++m) {
-    Cluster_index j = ds.find_set(d[2 * m]);
-    Cluster_index k = ds.find_set(d[2 * m + 1]);
+    Cluster_index j = ds.find_set(children(m, 0));
+    Cluster_index k = ds.find_set(children(m, 1));
     assert(j != k);
     ds.make_set(i);
     ds.link(i, j);
     ds.link(ds.find_set(i), k);
     ++i;
   }
   Cluster_index next_cluster_name = 0;
-  std::vector<Cluster_index> ret;
-  ret.reserve(n_leaves);
+  std::vector<Cluster_index>* ret = new std::vector<Cluster_index>();
+  ret->reserve(n_leaves);
   for (Cluster_index j = 0; j < n_leaves; ++j) {
     Cluster_index k = ds.find_set(j);
     if (ds_bas[k].name == -1) ds_bas[k].name = next_cluster_name++;
-    ret.push_back(ds_bas[k].name);
+    ret->push_back(ds_bas[k].name);
   }
-  return py::array(ret.size(), ret.data());
+  return nb::ndarray<Cluster_index, nanobind::numpy>(ret->data(), {ret->size()}, nb::capsule(ret, [](void* p) noexcept {
+                                                       delete reinterpret_cast<std::vector<Cluster_index>*>(p);
+                                                     }));
 }
 
 // TODO: Do a special version when ngb is a numpy array, where we can cast to int[k][n] ?
-// py::isinstance<py::array_t<std::int32_t>> (ou py::isinstance<py::array> et tester dtype) et flags&c_style
-// ou overload (en virant forcecast?)
-// aussi le faire au cas où on n'aurait pas un tableau, mais où chaque liste de voisins serait un tableau ?
-auto hierarchy(py::handle ngb, py::array_t<double, py::array::c_style | py::array::forcecast> density) {
+// also do this in the case where we don't have an array, but each list of neighbours is an array ?
+auto hierarchy(nb::object ngb, nb::ndarray<double, nb::ndim<1>, nb::c_contig> density)
```

Comment:
I wonder if c_contig does the right thing. What happens if I give Tomato a density thing that is not contiguous? With pybind11, I think it copied things to a contiguous array. I am afraid nanobind may reject the input.

Reply:
I don't remember the details, but `forcecast` was probably there for a reason. Looking at the pybind11 doc:
> The py::array::forcecast argument [...] ensures that non-conforming arguments are converted into an array satisfying the specified requirements instead of trying the next function overload.

Now I have no idea if we should try somehow to emulate this old behavior, or if the code would work just fine with any (not necessarily contiguous) array, that will require a closer look.